File: codes/python/chap06/answer_A39.py
# 입력
# A는 (종료 시각, 시작 시각)으로 되어 있음에 주의 [종료 시각의 오름차순으로 정렬하기 위함].
N = int(input())
A = []
for i in range(N):
	l, r = map(int, input().split())
	A.append([r, l])

# 정렬
A.sort()

# 종료 시각이 빠른 것부터 탐욕적으로 취해 간다(CurrentTime은 현재 시각)
CurrentTime = 0
Answer = 0
for i in range(N):
	if CurrentTime <= A[i][1]:
		CurrentTime = A[i][0]
		Answer += 1

# 출력
print(Answer)

File: codes/cpp/chap06/answer_A39.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int N, L[300009], R[300009];
vector<pair<int, int>> tmp; // 영화를 정렬하기 위한 임시 변수

int main() {
	// 입력
	cin >> N;
	for (int i = 1; i <= N; i++) {
		cin >> L[i] >> R[i];
		tmp.push_back(make_pair(R[i], L[i]));
	}

	// R의 오름차순으로 정렬
	sort(tmp.begin(), tmp.end());
	for (int i = 1; i <= N; i++) {
		R[i] = tmp[i - 1].first;
		L[i] = tmp[i - 1].second;
	}

	// 종료 시각이 빠른 것부터 탐욕적으로 취해 간다(CurrentTime은 현재 시각)
	int CurrentTime = 0, Answer = 0;
	for (int i = 1; i <= N; i++) {
		if (CurrentTime <= L[i]) {
			CurrentTime = R[i];
			Answer += 1;
		}
	}
	cout << Answer << endl;
	return 0;
}
